### PR TITLE
Added type name difference resolver for displaying differences in type

### PR DESCRIPTION
**Pull request by OmicronPersei**
Closes issue #2289 .

The feature adds the ability to display a shortened text output of the differences between two `Type`s when an equality of two objects is performed.  It also better handles comparing generic types.

### CODE REVIEW

File: src/NUnitFramework/tests/Constraints/EqualConstraintTests.cs [hunk 1]
```diff
@@ -666,6 +666,66 @@ public void SameValueDifferentTypeExactMessageMatch()
...
+                this.value = value;
+            }
+
+            public override string ToString()
+            {
+                return "Dummy " + value;
+            }
+        }
+
+        class DummyTemplatedClass<T>
```
Comment (jnm2) [inline src/NUnitFramework/tests/Constraints/EqualConstraintTests.cs]: Complete nit: the IL/C# lingo is "generic"– C++ has templates, IL and C# do not.

File: src/NUnitFramework/framework/Internal/TypeNameDifferenceResolverUtils.cs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        }
+
+        /// <summary>
+        /// Get the generic parameters of a given <see cref="Type"/>.
+        /// </summary>
+        public List<Type> GetGenericParams(Type type)
+        {
+            Guard.ArgumentNotNull(type, nameof(type));
+
+            return new List<Type>(type.GetGenericArguments());
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/TypeNameDifferenceResolverUtils.cs]: I'm curious why we don't use `type.GetGenericArguments()` instead of maintaining the helper method? Also arguments and parameters are not the same thing, though the reflection API is already abusing the distinction...

File: src/NUnitFramework/framework/Internal/TypeNameDifferenceResolverUtils.cs [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        }
+
+        /// <summary>
+        /// Shorten the given <see cref="Type"/> names by only including the relevant differing namespaces/types, if they differ.
+        /// </summary>
+        /// <param name="expectedType">The expected <see cref="Type"/>.</param>
+        /// <param name="actuallType">The actual <see cref="Type"/>.</param>
+        /// <param name="expectedTypeShortened">The shortened expected <see cref="Type"/> name.</param>
+        /// <param name="actualTypeShortened">The shortened actual <see cref="Type"/> name.</param>
+        public void ShortenTypeNames(Type expectedType, Type actuallType, out string expectedTypeShortened, out string actualTypeShortened)
```
Comment (jnm2) [inline src/NUnitFramework/framework/Internal/TypeNameDifferenceResolverUtils.cs]: `actuallType` should be `actualType`

File: src/NUnitFramework/tests/Constraints/EqualConstraintTests.cs (hunk 1 above)
Comment (OmicronPersei) [reply src/NUnitFramework/tests/Constraints/EqualConstraintTests.cs]: Whoops.  Will rename then.

File: src/NUnitFramework/framework/Internal/TypeNameDifferenceResolverUtils.cs (hunk 2 above)
Comment (OmicronPersei) [reply src/NUnitFramework/framework/Internal/TypeNameDifferenceResolverUtils.cs]: I think I intended to refactor that away after I refactored my functionality to use `Type`s instead of relying on the string representation of a `Type`. I think you're right, we don't need a new method for that, so as long as I put `using NUnit.Compatability;` at the top of the file!

**Comment by OmicronPersei** [comment]
> The utils class is bugging me. I can't tell what responsibility is being isolated into the utils class. Every helper method could live happily in the main class in my opinion. Am I overlooking something we'd lose by moving them in?

Well, I was worried it would violate SRP.  The utilities provided aren't necessarily to provide `TypeNameDifferenceResolver`, but are solely used by the class.  Truthfully, I'm new to getting my code to conform to things like SRP, maybe I was too eager to move it to its own class.

**Comment by jnm2** [comment]
> Well, I was worried it would violate SRP.

I don't think it matters in this case. In my mind to whatever extent you were violating SRP by having `IsGenericType` next to the methods in `TypeNameDifferenceResolver`, that's the same extent that you're still violating SRP by having it next to `TypeNameDifferenceResolverUtils.GetShortenedGenericTypes`.
If you were to worry about that, you'd add an `IsGenericType` extension method to `ReflectionExtensions`. **But** again I don't think it's worth worrying about. 

What I don't like is that `Foo` and `FooUtils` are acting as one class. By its name `FooUtils` is just an extension of `Foo`, each totally dependent on the other and pretty much useless without the other.

The type of scenario SRP should kick in for is if you had a bunch of related methods for which it would be reasonable to have more than one implementation, say for example parsing and formatting type names using a C#- or VB.NET-specific style rather than the generic CLR style names. Then instead of `Utils` you'd call the class `FormatProvider` or something like that. Formatting to a specific language in that scenario is a "reason to change" (because you might need to change to a new language) and so should be abstracted into a separate class through an interface or base class.
It's also usually unfruitful to guess ahead at potential future abstraction points so I'd wait until we have a concrete scenario. We aren't designing an API and we can quickly refactor at any time.

tl;dr; Since from what I see we don't have more than one "reason to change" or responsibility, let's stick with one class.

**Comment by jnm2** [comment]
Restarted the failing Travis job.

**Comment by mikkelbu** [comment]
Restarting CI once more

**Comment by jnm2** [comment]
> Restarting CI once more

Oh boy.

You can restart individual jobs rather than the whole build, but either way, the original build log disappears! I actually wanted to file and issue on the error I saw before I clicked Restart Job and now it's too late.

**Comment by mikkelbu** [comment]
> You can restart individual jobs rather than the whole build, but either way, the original build log disappears! 
I first discovered while restarting this time.

> I actually wanted to file and issue on the error I saw before I clicked Restart Job and now it's too late.
Sorry. We can hope for (yet) another error

**Comment by jnm2** [comment]
I'm sorry, the "oh boy" was directed at Travis, not you.

**Comment by mikkelbu** [comment]
> I'm sorry, the "oh boy" was directed at Travis, not you.

No need to be sorry 😄  (and I think that the build is going to fail again)
